design/open/lakefs-on-kv-testing-plan.md

```diff
@@ -0,0 +1,133 @@
+Updates are done for commitID, as part of `post` hooks
+* Currently unit tests cover hooks run and verify the results in DB using the service reading functions. There are also error testings (hook failures)
+* 67.6% coverage
+* What is missing:
+  * DB Failure during hook action (TBD - what is the expected behavior? Is that interesting to test?)
+  * Performance under load - not sure a load on actions/hooks is relevant
+  * Concurrent actions/hooks execution, which derives concurrent DB accesses. This gets more interesting if load is involved (in case load is relevant)
+
+### ```pkg/auth```
+* DB is used to store installation metadata, users and policies
+* Only 6.8% coverage
+* Some benchmarks exists that covers effectivePolicies (multi-table JOIN)
+* What is missing:
+  * Need to increase code coverage
+  * Comparison benchmark to toggle between Tables and KV and verify there is no degradation
+  * Migration tests:
+    * verify that data is migrated correctly
+    * Authentication transactions in during migration (data that was written to tables is read from KV)
+
+### ```pkg/diagnostics```
+* Only read access
+* No test coverage. Need to verify all is working the same after migration
+
+### ```pkg/graveler/ref```
+* Branch locking (read)
+* Repos, branches, commits and tags - read/write
+* 95.5% testing coverage
+* No performance/benchmarking
+* Missing:
+  * Migration tests
+  * Comparison benchmark
+
+### ```pkg/graveler/retention```
+* No direct DB access. Accesses are done using ```pkg/graveler/ref```, which should make the DB transition seamless
+* 42.9% coverage in unit testing
+* Missing performance tests that might help detect degradation
+
+### ```pkg/graveler/staging```
+* 89.1% coverage
+* Covers all staged KV functionality
+* No performance/benchmarking
+* Missing:
+  * Migration tests
+  * Comparison benchmark
+
+### Ref-Dump/Restore
+Currently this functionality is not covered directly, but since it relies on '''pkg/graveler/ref''' for DB access, the DB migration should be seamless
+It can be leveraged, however, to extend the cover of '''pkg/graveler/ref''' and for performance, as it is quite exhaustive (traverses all branches and commits and tags, per repo)
+
+## Execution Plan
+
+### KVM1
+* Data level migration tests infrastructure
+  * Migrate data from Table to KV, extract both and compare
+    * Data extraction should be done by listing all objects in the DB, using a designated 'get' function, and compare
+  * Implement ~~dumpers for `gateway_multiprts`~~ `GetAll` for `multiparts.Tracker`, to return a list of `MultipartUpload`
+  * Implement comparison of `MultipartUploads` list. Lists are considered identical if objects are identical, but **not necessarily** at the same order
+  * Implement unit tests for `pkg/gateway/multiparts`
+    * Add multipart uploads using `multiparts.Tracker.Create` with Table DB (KV Feature Flag off)
```

Comment:
As I commented before - I think the migration tests should rely on reading data from a file and not to create the data dynamically - the migration process should happen in the lakeFS entry point before server is up

Reply:
I think the migration tests should rely on the migration method that will be implemented - if it is dump&read or any other method.
Either ways, I agree with @itaiad200 that we should use the store library functions to read the data to compare. As data changes and grows, the library will obviously adapt, as to support the modified data. I'm not sure the same goes for dump support, as even if the initial implementation of the migration relies on dump, and so, will be maintained with any schema change, we might decide to replace this method with a different one, leaving us with the need to keep and support the dump functionality for the sole purpose of testing